pkg/sdn/plugin/sdn-cni-plugin/openshift-sdn.go

```diff
@@ -0,0 +1,511 @@
+// +build linux
+
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"net"
+	"os"
+	"os/exec"
+	"strconv"
+	"strings"
+	"syscall"
+
+	sdnplugin "github.com/openshift/origin/pkg/sdn/plugin"
+	"github.com/openshift/origin/pkg/sdn/plugin/api"
+	"github.com/openshift/origin/pkg/util/netutils"
+
+	oclient "github.com/openshift/origin/pkg/client"
+	configapi "github.com/openshift/origin/pkg/cmd/server/api"
+	osapi "github.com/openshift/origin/pkg/sdn/api"
+
+	kclient "k8s.io/kubernetes/pkg/client/unversioned"
+	knetwork "k8s.io/kubernetes/pkg/kubelet/network"
+	kbandwidth "k8s.io/kubernetes/pkg/util/bandwidth"
+
+	"github.com/containernetworking/cni/pkg/ip"
+	"github.com/containernetworking/cni/pkg/ipam"
+	"github.com/containernetworking/cni/pkg/ns"
+	"github.com/containernetworking/cni/pkg/skel"
+	"github.com/containernetworking/cni/pkg/types"
+
+	"github.com/vishvananda/netlink"
+
+	// Must be imported to register API types like ClusterNetwork
+	_ "github.com/openshift/origin/pkg/api/install"
+)
+
+const (
+	sdnScript   = "openshift-sdn-ovs"
+	setUpCmd    = "setup"
+	tearDownCmd = "teardown"
+	statusCmd   = "status"
+	updateCmd   = "update"
+
+	AssignMacVlanAnnotation string = "pod.network.openshift.io/assign-macvlan"
+
+	interfaceName = knetwork.DefaultInterfaceName
+)
+
+type PodConfig struct {
+	clusterNetwork   string
+	nodeNetwork      string
+	mtu              uint32
+	vnid             string
+	ingressBandwidth string
+	egressBandwidth  string
+	wantMacvlan      bool
+}
+
+type podInfo struct {
+	Vnid        uint32
+	Privileged  bool
+	Annotations map[string]string
+}
+
+func gatherCniArgs(cniArgs string) (map[string]string, error) {
+	mapArgs := make(map[string]string)
+	for _, arg := range strings.Split(cniArgs, ";") {
+		parts := strings.Split(arg, "=")
+		if len(parts) != 2 {
+			return nil, fmt.Errorf("invalid CNI_ARG '%s'", arg)
+		}
+		mapArgs[strings.TrimSpace(parts[0])] = strings.TrimSpace(parts[1])
+	}
+	return mapArgs, nil
+}
+
+func readPodInfo(originClient *oclient.Client, kubeClient *kclient.Client, cniArgs string, pluginName string) (*podInfo, error) {
+	mapArgs, err := gatherCniArgs(cniArgs)
+	if err != nil {
+		return nil, err
+	}
+
+	namespace, namespaceOk := mapArgs["K8S_POD_NAMESPACE"]
+	name, nameOk := mapArgs["K8S_POD_NAME"]
+	if !namespaceOk || !nameOk {
+		return nil, fmt.Errorf("missing pod namespace or name")
+	}
+
+	info := &podInfo{}
+
+	if sdnplugin.IsOpenShiftMultitenantNetworkPlugin(pluginName) {
+		netNamespace, err := originClient.NetNamespaces().Get(namespace)
+		if err != nil {
+			return nil, err
+		}
+		info.Vnid = netNamespace.NetID
+	}
+
+	// FIXME: does this ensure the returned pod lives on this node?
+	pod, err := kubeClient.Pods(namespace).Get(name)
```

Comment:
it doesn't, but if it's not, then it's a bug in somebody else's code that we even got called...